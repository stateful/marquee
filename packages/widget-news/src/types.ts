import { HN_CHANNELS } from './constants'

export interface HackerNews {
  comments_count: number
  domain: string
  id: number
  points: number
  time: number
  time_ago: string
  title: string
  type: string
  url: string
  user: string
}

export type HackerNewsChannels = typeof HN_CHANNELS[number]

export interface WidgetState {
  news: HackerNews[]
  isFetching: boolean,
  channel: HackerNewsChannels,
  error?: Error
}
