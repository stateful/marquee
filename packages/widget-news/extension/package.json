{
  "name": "@vscode-marquee/widget-news-extension",
  "description": "Extension logic for News Widget",
  "main": "../build/extension.js",
  "module": "../build/extension.js",
  "types": "../build/extension.d.ts"
}
