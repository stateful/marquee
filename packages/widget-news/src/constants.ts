import { State, Configuration } from './types'

export const MIN_UPDATE_INTERVAL = 1000 * 60
export const FETCH_DATA_TIMEOUT = 10000
export const DEFAULT_STATE: State = {
  news: [],
  isFetching: true,
  channel: 'HN Frontpage',
  error: null
}
export const DEFAULT_CONFIGURATION: Configuration = {
  feeds: {},
  updateInterval: MIN_UPDATE_INTERVAL
}
