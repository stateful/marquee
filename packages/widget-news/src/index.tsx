import React from 'react';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faNewspaper } from "@fortawesome/free-solid-svg-icons";

import News from "./Widget";

export default {
  name: "news",
  icon: <FontAwesomeIcon icon={faNewspaper} />,
  label: 'News',
  tags: ["hackernews", "HNN", "programing", "entertainment"],
  description: "The Hacker News feed, with more feeds coming soon.",
  component: News,
};
export { News };
