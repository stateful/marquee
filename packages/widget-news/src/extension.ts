import vscode from 'vscode'
import Parser from 'rss-parser'
import ExtensionManager from '@vscode-marquee/utils/extension'

import { DEFAULT_CONFIGURATION, DEFAULT_STATE } from './constants'
import type { Configuration, FeedItem, State } from './types'

const STATE_KEY = 'widgets.news'
export class NewsExtensionManager extends ExtensionManager<State, Configuration> {
  private _parser = new Parser()
  private _isFetching = false

  constructor (context: vscode.ExtensionContext, channel: vscode.OutputChannel) {
    super(context, channel, STATE_KEY, DEFAULT_CONFIGURATION, DEFAULT_STATE)
    this.fetchFeeds()
    this.on('stateUpdate', () => this.fetchFeeds())
    setInterval(() => this.fetchFeeds(), this.configuration.updateInterval)
  }

  async fetchFeeds () {
    if (this._isFetching) {
      return
    }

    this._isFetching = true
    await this.updateState('isFetching', true)

    try {
      const url = this._configuration.feeds[this._state.channel]
      if (!url) {
        throw new Error(
          `Channel "${this._state.channel}" not found, ` +
          `available channels are ${Object.keys(this._configuration.feeds).join(', ')}`
        )
      }

      this._channel.appendLine(`Fetch News ("${this._state.channel}") from ${url}`)
      const feed = await this._parser.parseURL(url)

      await this.updateState('news', feed.items as FeedItem[])
      await this.updateState('isFetching', false)
      await this.updateState('error', null)
      this._tangle?.broadcast({
        news: feed.items,
        isFetching: false,
        error: null
      } as State & Configuration)
      setTimeout(() => { this._isFetching = false }, 100)
    } catch (err: any) {
      await this.updateState('isFetching', false)
      await this.updateState('error', err.message as string)
      this._tangle?.broadcast({
        news: [] as FeedItem[],
        isFetching: false,
        error: err.message
      } as State & Configuration)

      setTimeout(() => {
        this._tangle?.broadcast({ isFetching: false } as State & Configuration)
        this._isFetching = false
      }, 100)
    }
  }
}

export function activate (
  context: vscode.ExtensionContext,
  channel: vscode.OutputChannel
) {
  const stateManager = new NewsExtensionManager(context, channel)
  return {
    marquee: {
      disposable: stateManager,
      defaultState: stateManager.state,
      defaultConfiguration: stateManager.configuration,
      setup: stateManager.setBroadcaster.bind(stateManager)
    },
  }
}
