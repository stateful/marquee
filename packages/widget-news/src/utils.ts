import { HN_API } from './constants'
import type { WidgetState, HackerNews } from './types'

export async function fetchNews (data: WidgetState) {
  const resp = await fetch(
    HN_API + data.channel,
    { mode: 'cors' }
  ).catch((err) => ({
    ok: false,
    status: 0,
    json: () => err
  }))

  if (!resp.ok) {
    return {
      ...data,
      isFetching: false,
      error: new Error(`Failed to fetch news! (status: ${resp.status})`)
    }
  }

  return {
    isFetching: false,
    news: await resp.json() as HackerNews[],
    channel: data.channel
  }
}
