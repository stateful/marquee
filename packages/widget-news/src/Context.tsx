import React, { createContext } from 'react'
import { connect, getEventListener, MarqueeWindow } from '@vscode-marquee/utils'

import type { Context, Configuration, State } from './types'

declare const window: MarqueeWindow
const NewsContext = createContext<Context>({} as Context)
export const WIDGET_ID = '@vscode-marquee/news-widget'

export const NewsProvider = ({ children }: { children: React.ReactElement }) => {
  const widgetState = getEventListener<Configuration & State>(WIDGET_ID)
  const providerValues = connect<Configuration & State>({
    ...window.marqueeStateConfiguration[WIDGET_ID].configuration,
    ...window.marqueeStateConfiguration[WIDGET_ID].state
  }, widgetState)

  return (
    <NewsContext.Provider value={{ ...providerValues }}>
      {children}
    </NewsContext.Provider>
  )
}

export default NewsContext
