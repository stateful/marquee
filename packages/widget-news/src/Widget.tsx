import React, { useState, useEffect } from 'react'
import {
  Grid,
  Link,
  List,
  ListItem,
  ListItemText,
  ListItemAvatar,
  Box,
} from '@mui/material'
import Typography from '@mui/material/Typography'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faHackerNews } from '@fortawesome/free-brands-svg-icons/faHackerNews'
import CircularProgress from '@mui/material/CircularProgress'

import wrapper, { Dragger } from '@vscode-marquee/widget'
import { NetworkError } from '@vscode-marquee/utils'

import PopMenu from './components/Pop'
import { fetchNews } from './utils'
import { DEFAULT_STATE } from './constants'
import type { WidgetState } from './types'

let News = () => {
  const [data, setData] = useState(DEFAULT_STATE)
  useEffect(() => {
    let _setData = (data: WidgetState) => setData(data)
    setData({ ...data, isFetching: true })
    fetchNews(data).then((data) => _setData(data))
    return () => { _setData = () => {} }
  }, [data.channel])

  return (
    <>
      <Grid item xs={1} style={{ maxWidth: '100%' }}>
        <Box sx={{
          borderBottom: '1px solid var(--vscode-foreground)',
          padding: '8px',
        }}>
          <Grid
            container
            direction="row"
            justifyContent="space-between"
            alignItems="center"
          >
            <Grid item>
              <Typography variant="subtitle1">News</Typography>
            </Grid>
            <Grid item>
              <Grid container direction="row" spacing={1}>
                <Grid item>
                  <PopMenu value={data.channel} onChannelChange={(channel) => setData({ ...data, channel })} />
                </Grid>
                <Grid item>
                  <Dragger />
                </Grid>
              </Grid>
            </Grid>
          </Grid>
        </Box>
      </Grid>
      <Grid item xs>
        <Grid
          container
          wrap="nowrap"
          direction="column"
          style={{ height: '100%' }}
        >
          <Grid item xs style={{ overflow: 'auto' }}>
            {data.error && (
              <Grid
                item
                xs
                style={{
                  overflow: 'auto',
                  height: '100%',
                  width: '100%',
                  padding: '24px',
                }}
              >
                <NetworkError message={data.error.message} />
              </Grid>
            )}
            {data.isFetching && (
              <Grid
                container
                style={{ height: '100%' }}
                alignItems="center"
                justifyContent="center"
                direction="column"
              >
                <Grid item>
                  <CircularProgress color="secondary" />
                </Grid>
              </Grid>
            )}
            {!data.isFetching && data.news.length === 0 && (
              <Grid
                container
                style={{ height: '100%' }}
                alignItems="center"
                justifyContent="center"
                direction="column"
              >
                <Grid item>
                  No news available at the moment!
                </Grid>
              </Grid>
            )}
            {!data.isFetching && data.news.length !== 0 && (
              <List dense={true}>
                {data.news.map((entry) => (
                  <ListItem dense key={entry.id}>
                    <ListItemAvatar>
                      <Grid container justifyContent="center" alignItems="center">
                        <Grid item>
                          <FontAwesomeIcon icon={faHackerNews} />
                        </Grid>
                      </Grid>
                    </ListItemAvatar>
                    <ListItemText
                      primary={
                        <>
                          <Link
                            component="a"
                            href={`https://news.ycombinator.com/item?id=${entry.id}`}
                            target="_blank"
                            underline="hover">
                            {entry.title}
                          </Link>
                          {entry.domain && (
                            <Typography variant="caption">
                              &nbsp;(
                              <Link
                                component="a"
                                href={entry.url}
                                target="_blank"
                                underline="hover"
                              >
                                {entry.domain}
                              </Link>
                              )
                            </Typography>
                          )}
                        </>
                      }
                      secondary={
                        <Typography style={{ fontSize: '10px' }}>
                          {entry.points} points by {entry.user} &nbsp;
                          {entry.time_ago}
                        </Typography>
                      }
                    />
                  </ListItem>
                ))}
              </List>
            )}
          </Grid>
        </Grid>
      </Grid>
    </>
  )
}

export default wrapper(News, 'news')
