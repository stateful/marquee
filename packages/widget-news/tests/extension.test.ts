import { activate } from '../src/extension'

describe('Extension Manager', () => {
  let manager: any

  beforeEach(() => {
    manager = activate({} as any)
    ;(manager.marquee.disposable.updateState as jest.Mock).mockClear()
  })

  it('does nothing if _isFetching is set to true', async () => {
    ;(manager.marquee.disposable.updateState as jest.Mock).mockClear()
    await manager.marquee.disposable.fetchFeeds()
    expect(manager.marquee.disposable.updateState).toBeCalledTimes(0)
  })

  it('throws an error if channel wasn\'t found', async () => {
    (manager.marquee.disposable.updateState as jest.Mock).mockClear()
    manager.marquee.disposable['_isFetching'] = false
    const tangle = { broadcast: jest.fn() }

    manager.marquee.disposable._tangle = tangle
    manager.marquee.disposable._state = { channel: 'foobar' }
    manager.marquee.disposable._configuration = {
      feeds: {
        barfoo: 'some url',
        somethingElse: 'some url'
      }
    }
    await manager.marquee.disposable.fetchFeeds()
    expect(tangle.broadcast.mock.calls).toMatchSnapshot()
    expect(manager.marquee.disposable.updateState.mock.calls)
      .toMatchSnapshot()
  })

  it('can successfully fetch items', async () => {
    (manager.marquee.disposable.updateState as jest.Mock).mockClear()
    manager.marquee.disposable['_isFetching'] = false
    const tangle = { broadcast: jest.fn() }

    manager.marquee.disposable._tangle = tangle
    manager.marquee.disposable._state = { channel: 'barfoo' }
    manager.marquee.disposable._configuration = {
      feeds: {
        barfoo: 'some url',
        somethingElse: 'some url'
      }
    }
    await manager.marquee.disposable.fetchFeeds()
    expect(manager.marquee.disposable['_parser'].parseURL)
      .toBeCalledWith('some url')
    expect(tangle.broadcast.mock.calls).toMatchSnapshot()
    expect(manager.marquee.disposable.updateState.mock.calls)
      .toMatchSnapshot()
  })
})
