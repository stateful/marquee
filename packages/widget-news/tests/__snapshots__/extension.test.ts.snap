// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extension Manager can successfully fetch items 1`] = `
Array [
  Array [
    Object {
      "error": null,
      "isFetching": false,
      "news": Array [
        Object {
          "title": "rss item 1",
        },
        Object {
          "title": "rss item 2",
        },
      ],
    },
  ],
]
`;

exports[`Extension Manager can successfully fetch items 2`] = `
Array [
  Array [
    "isFetching",
    true,
  ],
  Array [
    "news",
    undefined,
  ],
  Array [
    "isFetching",
    false,
  ],
  Array [
    "error",
    null,
  ],
]
`;

exports[`Extension Manager throws an error if channel wasn't found 1`] = `
Array [
  Array [
    Object {
      "error": "Channel \\"foobar\\" to found, available channels are barfoo, somethingElse",
      "isFetching": false,
      "news": Array [],
    },
  ],
]
`;

exports[`Extension Manager throws an error if channel wasn't found 2`] = `
Array [
  Array [
    "isFetching",
    true,
  ],
  Array [
    "isFetching",
    false,
  ],
  Array [
    "error",
    "Channel \\"foobar\\" to found, available channels are barfoo, somethingElse",
  ],
]
`;
